### PR TITLE
Add zstd for gpfdist

### PR DESCRIPTION
**Pull request by zhaorui20005**
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/misc/fstream/gfile.c [hunk 1]
```diff
@@ -511,7 +511,256 @@ gz_file_open(gfile_t *fd)
...
+		size_t	ret;
+		/*
+		 * 'out->dst' is our output buffer.
+		 * 'pos' is a pointer to the next pos in 'out_buffer'
+		 * 'out_size' is num bytes currently read out of 'out_buffer'
+		 *
+		 * if s is >0 we have data in 'out.dst' that we didn't write
+		 * yet, write it and return.
+		 */
+		int s = zstd->out.pos - zstd->out_size;
```
Comment (fanfuxiaoran) [inline src/backend/utils/misc/fstream/gfile.c:546]: zstd->out.pos type is pointer or int?

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:546]: The structure definition is like this:
```
typedef struct ZSTD_inBuffer_s {
  const void* src;    /**< start of input buffer */
  size_t size;        /**< size of input buffer */
  size_t pos;         /**< position where reading stopped. Will be updated. Necessarily 0 <= pos <= size */
} ZSTD_inBuffer;

typedef struct ZSTD_outBuffer_s {
  void*  dst;         /**< start of output buffer */
  size_t size;        /**< size of output buffer */
  size_t pos;         /**< position where writing stopped. Will be updated. Necessarily 0 <= pos <= size */
} ZSTD_outBuffer;```

Comment (fanfuxiaoran) [reply src/backend/utils/misc/fstream/gfile.c:546]: Ok, but the comment on line  540 is a little confusing, maybe not correct.

File: src/backend/utils/misc/fstream/gfile.c [hunk 2]
```diff
@@ -511,7 +511,256 @@ gz_file_open(gfile_t *fd)
...
+	int in_size, out_size;
+	ZSTD_CStream* cstream;
+	ZSTD_DStream* dstream;
+	unsigned char in_buffer[COMPRESSION_BUFFER_SIZE];
+	unsigned char out_buffer[COMPRESSION_BUFFER_SIZE];
+};
+enum compress_mode
+{
+	ZSTD_CONTINUE_FLUSH = 0,
+	ZSTD_END_FLUSH = 2,
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/gfile.c:528]: why is it from 0 to 2? what is 1 in enum?

File: src/backend/utils/misc/fstream/gfile.c (hunk 1 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:546]: Right, the comments are copied from gz, I will fix it later. Thanks!

File: src/backend/utils/misc/fstream/gfile.c (hunk 2 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:528]: They are defined in codes in zstd.h in the version 1.4.8
```
/*===== Streaming compression functions =====*/
typedef enum {
    ZSTD_e_continue=0, /* collect more data, encoder decides when to output compressed result, for optimal compression ratio */
    ZSTD_e_flush=1,    /* flush any data provided so far,
                        * it creates (at least) one new block, that can be decoded immediately on reception;
                        * frame will continue: any future data can still reference previously compressed data, improving compression.
                        * note : multithreaded compression will block to flush as much output as possible. */
    ZSTD_e_end=2       /* flush any remaining data _and_ close current frame.
                        * note that frame is only closed after compressed data is fully flushed (return value == 0).
                        * After that point, any additional data starts a new frame.
                        * note : each frame is independent (does not reference any content from previous frame).
                        : note : multithreaded compression will block to flush as much output as possible. */
} ZSTD_EndDirective;
```
But the zstd.h in pipelines compiling version is 1.3.7, and the enum values are not defined, so I borrow this enum from later version and use only the first one and the last one.
Should I add ZSTD_e_flush too?

File: src/backend/utils/misc/fstream/gfile.c [hunk 3]
```diff
@@ -511,7 +511,256 @@ gz_file_open(gfile_t *fd)
...
+	{
+		/*
+		 * writing a compressed file
+		 */
+		if (!(fd->u.zstd->cstream = ZSTD_createCStream()))
+		{
+			gfile_printf_then_putc_newline("ZSTD_createCStream() failed");
+			return 1;
+		}
+		if(ZSTD_isError(ZSTD_initCStream(fd->u.zstd->cstream, 3)))
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/gfile.c:754]: it is better to use the enum name in zstd.h instead of hard code `3` for compression strategies

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:754]: OK, I have found the macro ZSTD_CLEVEL_DEFAULT which is substitute of 3 and which is defined in both version of zstd.h

File: src/backend/utils/misc/fstream/gfile.c [hunk 4]
```diff
@@ -511,7 +511,256 @@ gz_file_open(gfile_t *fd)
...
+	int ret;
+	size_t left_to_compress = size;
+	size_t one_iter_compress;
+	struct zstdlib_stuff* zstd = fd->u.zstd;
+
+	do
+	{
+		/*
+		 * we do not wish that the size of the input buffer to the ZSTD_compressStream2 engine, will be greater
+		 * than the recomended COMPRESSION_BUFFER_SIZE.
```
Comment (huiliang-liu) [inline src/backend/utils/misc/fstream/gfile.c:668]: recommended

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:668]: I shall check all the comments and fix them later.Thanks!

File: src/backend/utils/misc/fstream/gfile.c [hunk 5]
```diff
@@ -511,7 +511,256 @@ gz_file_open(gfile_t *fd)
...
+{
+	if (!(fd->u.zstd = gfile_malloc(sizeof *fd->u.zstd)))
+	{
+		gfile_printf_then_putc_newline("Out of memory");
+		return 1;
+	}
+
+	memset(fd->u.zstd, 0, sizeof *fd->u.zstd);
+
+	fd->u.zstd->in.src = fd->u.zstd->in_buffer;
```
Comment (fanfuxiaoran) [inline src/backend/utils/misc/fstream/gfile.c:721]: Personally, I think the fileds 
```
int in_size, out_size;
unsigned char in_buffer[COMPRESSION_BUFFER_SIZE];
unsigned char out_buffer[COMPRESSION_BUFFER_SIZE];"
```
They are usless. Instead we can use zstd->in and zstd->out directly, they are really confusing.

File: src/backend/utils/misc/fstream/gfile.c (hunk 2 above)
Comment (huiliang-liu) [reply src/backend/utils/misc/fstream/gfile.c:528]: I think we can add it if there is no build warning, or add it in comment.

File: src/backend/utils/misc/fstream/gfile.c (hunk 5 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:721]: they are initialized in zstd_file_open with the following code:
```
	fd->u.zstd->in.src = fd->u.zstd->in_buffer;
	fd->u.zstd->out.dst = fd->u.zstd->out_buffer;
```
So, here we use the fixed length of array to simplify the memory allocation process.

File: src/backend/utils/misc/fstream/gfile.c (hunk 2 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:528]: Ok. I will add it in the next commit.

File: src/backend/utils/misc/fstream/gfile.c (hunk 5 above)
Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:721]: The in_size and out_size are position that we have read or written, the pos in in or out are inner structure that could be changed be compression or decompression function.

Comment (fanfuxiaoran) [reply src/backend/utils/misc/fstream/gfile.c:721]:
```
 zstd->in.size = zstd->in_size;
```
and 
```
 if (zstd->in.pos == zstd->in_size)
```
The zstd->in.size and  and  zstd->in_size look same.

Comment (zhaorui20005) [reply src/backend/utils/misc/fstream/gfile.c:721]: Right! 
The "zstd->in_size" is the same to the "zstd->in,size" in this situation.
but the "zstd->in.size" is an inner structure that we should not change them too often. And the "zstd->in_size" is the pos to the data we have read into the in_buffer to be decompressed later.
Before we call decompress function, the 'zstd->in.size' should be initialized. 
By using the 'zstd->in_size' we can refer to the compress process 'zstd->out_size'.
And the decompression codes are modified from gz code. The most processes are the same, the differences are in details like structure type or APIs.

**Review by huiliang-liu — APPROVED** [review]

**Pull request merged**